### PR TITLE
gh-102859: Remove JUMP_IF_FALSE_OR_POP and JUMP_IF_TRUE_OR_POP

### LINKED ISSUE #102859: Remove JUMP_IF_FALSE_OR_POP and JUMP_IF_TRUE_OR_POP
See https://github.com/faster-cpython/ideas/issues/567.

### Linked PRs
* gh-102870

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #102859 .

* Issue: gh-102859

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -9105,21 +9091,30 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
...
+            opcode = inst->i_opcode;
+            oparg = inst->i_oparg;
+            if (HAS_TARGET(opcode)) {
+                assert(inst->i_target->b_iused > 0);
+                target = &inst->i_target->b_instr[0];
+                assert(!IS_ASSEMBLER_OPCODE(target->i_opcode));
+            }
+            else {
+                target = &nop;
+            }
```
Comment (carljm) [inline Python/compile.c:9151]: This seems redundant; `target` was already initialized to `&nop` above and nothing besides the `if` above will set it.

**Review by carljm — APPROVED** [review]

File: Python/compile.c (hunk 1 above)
Comment (iritkatriel) [reply Python/compile.c:9151]: I think we want to reset it for each iteration (even though it won’t be used when HAS_TARGET is false, it’s tidier.)

**Comment by iritkatriel** [comment]
Thanks, I’ll wait for benchmark results before merging.

File: Python/compile.c (hunk 1 above)
Comment (carljm) [reply Python/compile.c:9151]: Oops, I just missed that the initialization was outside the loop :) Looks good.

**Comment by markshannon** [comment]
https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20230322-3.12.0a6+-b2bf829

**Review by markshannon — APPROVED** [review]
I would expect this to cause a slowdown, but it seems the slowdown is negligible.

I expect that we will be able to get all that back and more with the optimizations discussed in the issue.

**Pull request merged**